### ISSUE TITLE
SCIM support

### ISSUE THREAD

**Issue by jan-peter-alten-bmw** [keycloak/keycloak#13484]
### Description

Looks like [KEYCLOAK-2537](https://issues.redhat.com/browse/KEYCLOAK-2537) did not yet got carried over here.

Keycloak should support at least [System for Cross-domain Identity Management](http://www.simplecloud.info/#Overview)

### Discussion

https://issues.redhat.com/browse/KEYCLOAK-2537

### Motivation

When using Keycloak as an identity provider group memberships can either be forwarded during SAML or OIDC logins to the service provider or an external LDAP.

If the service uses access tokens to grant basic access authentication, e.g. for command line tools, updates in group membership will not be updated until the next web based login of the user.

If Keycloak would support SCIM these changes could be forwarded to the service without the need to operate and maintain an LDAP server.

**Comment by uesleilima** [comment]
Interesting starting point: https://github.com/Captain-P-Goldfish/scim-for-keycloak

**Comment by InfoSec812** [comment]
@pierreozoux your implementation looks interesting, but the license (AGPL) would not be compatible with Keycloak (Apache 2.0). Would you and the authors be willing to work on a changing the license?

**Comment by pierreozoux** [comment]
@InfoSec812 if we can work together to merge it upstream, yes of course!

How can we discuss about this?
 - Async in this issue
 - maybe you have a monthly contributor call?
 - or maybe we can plan an ad hoc call for this topic?

If you could let us know, this would be great, we will have time in the coming months to work on this topic.

**Comment by gamunu** [comment]
I'm interested to see this merged upstream as well. 
@pierreozoux I saw that you had created a pull request to change the license to Apache 2. Is this something we can expect in the future?

**Comment by hrenard** [comment]
The license is now Apache 2. We'll resume work on the extension this year, hopefully during Q2, and we intend to put it in production in-house. Our goal is to make it simple to use SCIM in a collection of FOSS collaborative apps. Most features (that we require) are working.

We'd love to see the same features upstream, but as the main developer, I'm not sure if there is a path. Here is what I think could be blocking :
- We support only one SCIM simple topology that makes sense for us.
- The scim client reacts to events, so operations are not atomics.
- User federation is used for configuration but isn't used as intended (by Keycloak).
- We use a [scim library](https://github.com/pingidentity/scim2) which needs some dependencies.
- I'm not a very good java developer, so the code isn't as beautiful as it should.

Now, if a Keycloak developer can have a look and think there is a way to merge our efforts. I'll be glad to help.

**Comment by garpinc** [comment]
> We started some work to make it a scim client: https://lab.libreho.st/libre.sh/scim/keycloak-scim
> 
> And would be interested to work with the team to upstream this code.

This is client right? Wouldn't it be better to have a separate issue for scim server vs client?

**Comment by thomasjcf29** [comment]
Hey Guys, any update on this one? It would be really nice to get builtin SCIM Client support for Keycloak. This would allow Keycloak to automatically create, update, and delete users in third party services, like AWS.

**Comment by jonkoops** [comment]
@stianst This is an issue with quite a lot of upvotes, it might be worth it to take a look if this is something we want to implement.

**Comment by FrancescoErmini** [comment]
Hey guys, follow-up on this topic, any updates or plan for 2024?

**Comment by noraab** [comment]
Hi @stianst , from my point of view the (for me) most important feature would be to federate users from cloud directories like e.g. Microsoft Entra ID.

At the moment, the only option is to use LDAP which requires additional setup and additional services on the Microsoft Azure side. Also, LDAP doesn't support the "offboarding" of users.

**Comment by thomasjcf29** [comment]
It would also be useful for implementations where Keycloak is the source of truth for users (or has got the users by LDAP) to federate those users to other SCIM Servers to enable automatic provisioning and deprovisioning of users.

For example:
https://docs.aws.amazon.com/singlesignon/latest/userguide/provision-automatically.html

It would be useful for Keycloak to be able to automatically provision users through SCIM where needed for integration into services like AWS IAM Identity Control (IAM User replacement).

**Comment by blarghmatey** [comment]
I think that implementation of both server and client SCIM functionality as part of Keycloak core would add substantial value to the project. Numerous applications and services offer support for SCIM (both for provisioning users and their group information _into_ their platform, as well as populating an external identity system - Keycloak - with that information). There is currrently a third-party plugin that has implemented the server functionality and is working on client functionality (https://scim-for-keycloak.de/), but I also think that incorporating that protocol into Keycloak core would be immensely useful.

For reference, my team has forked and updated the plugin noted at the start of this conversation such that it currently works with Keycloak 23.x https://github.com/mitodl/keycloak-scim/

**Comment by thomasdarimont** [comment]
Great discussion so far! I'm wondering how you would test the SCIM compliance with the extensions mentioned in this discussion.

I just found this, which looks like an interesting start:
https://github.com/wso2-incubator/scim2-compliance-test-suite

**Comment by Hokwang** [comment]
I had same usage, I want to use keycloak with AWS IAM Identity Center (old AWS SSO).

**Comment by maxvisser** [comment]
We removed keycloak from our stack and now use azure ad (again) for our provisioning/sso because this was a crucial feature missing without installing plugings.

**Comment by jonkoops** [comment]
FYI if someone has a good idea of an what an implementation of SCIM should look like we're more than willing to accept proposals and PRs for developing this as an experimental feature. This issue is on the backlog of the team, but we have more immediate priorities to take care of before we can start getting to this.

**Comment by jochenr** [comment]
I cannot tell you how the implementation should look like, but I can tell you a good use case.

Using github.com Enterprise with Keycloak as described in this documentation

- https://docs.github.com/en/enterprise-cloud@latest/admin/identity-and-access-management/using-saml-for-enterprise-iam/configuring-saml-single-sign-on-for-your-enterprise
- so that it works as with okta:
https://docs.github.com/en/enterprise-cloud@latest/admin/identity-and-access-management/provisioning-user-accounts-for-enterprise-managed-users/configuring-scim-provisioning-with-okta

would be great.


PS: Actually it would even be better if github.com supports OIDC / OAuth2 with Keycloak, but surprisingly they only support OIDC with EntraID😉

**Comment by jdhakar1995** [comment]
> I think that implementation of both server and client SCIM functionality as part of Keycloak core would add substantial value to the project. Numerous applications and services offer support for SCIM (both for provisioning users and their group information _into_ their platform, as well as populating an external identity system - Keycloak - with that information). There is currrently a third-party plugin that has implemented the server functionality and is working on client functionality (https://scim-for-keycloak.de/), but I also think that incorporating that protocol into Keycloak core would be immensely useful.
> 
> For reference, my team has forked and updated the plugin noted at the start of this conversation such that it currently works with Keycloak 23.x https://github.com/mitodl/keycloak-scim/

I want to sync users and groups from keycloak to AWS IAM Identity Center via SCIM. Does this plugin support it?

**Comment by randomstuff** [comment]
Would that make sense (less unstable wrt API break) to have such functionality implemented a separate application (communicating with the Keycloak API) instead of a plugin? (Though it would probably be better to have it included as part of Keycloak itself)

**Comment by alexsedlex** [comment]
Hi everyone ! With his supervision, we leveraged on the @hrenard SCIM extension for keycloack. This extension now allows to propagate user/group modifications made in Keycloack to all registered SCIM endpoints, and to synchronize users/groups from registered SCIM endpoints. 

The documentation & code is available [here](https://lab.libreho.st/libre.sh/scim/keycloak-scim) and waiting for your feedbacks !
We tested it by wiring a Nextcloud instance supporting SCIM protocol, but feel free to test it on your own SCIM tools.

We think it would make sense to upstream this extension, as more and more tools provide SCIM APIs. Could someone from the core dev team guide us through the PR process ?
Thanks a lot !

**Comment by jonkoops** [comment]
Hi @alexsedlex, we would more than welcome a community contribution on this matter. I would recommend that you read our [contribution guidelines](https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md) to get started.

The best approach would be to start working on this feature in incremental small pull requests behind a [feature profile](https://github.com/keycloak/keycloak/blob/main/common/src/main/java/org/keycloak/common/Profile.java). This reduces the review load as small changes are easy to review, and allows you to get early feedback from the community through nightly builds and new releases behind an experimental flag.

**Comment by sventorben** [comment]
First of all, thank you for the great work – this feature looks really promising and would be a big step forward for Keycloak!

I have two questions regarding this:

1. If I remember correctly, the "Organization Feature" had a to-do item for implementing SCIM functionalities. Could you explain how that relates to this feature?
2. When using Keycloak Event Listeners here, would this work in conjunction with LDAP as well? From my understanding, no events are triggered in Keycloak during an LDAP synchronization. Would that be different in this case?

Thanks in advance for your response and for the awesome work!

**Comment by alexsedlex** [comment]
Hi everyone ! Following your advices we've made a first [draft PR,](https://github.com/keycloak/keycloak/pull/34733) make our best to follow the guidelines. Please let us know what you think and feel free to discuss and modify ! 

I'm still struggling on an Hibernate issue : the modules provides a new hibernate entity (ScimResourceMapping) through the JpaEntityProviderFactory mechanism. It works perfectly if the SCIM extension is declared on its own (see [original implementation](https://forge.libre.sh/libre.sh/keycloak-scim) ), but not if the module is included in the main codebase.
One way to fix this would be to declared this entity with the other default entities in jpa.entities and default-persistence.xml, but as this feature is still experimental I think it makes more sense to make it optional and use the JpaEntityProviderFactory mechanism.
Any help much appreciated :)

**Comment by kochen** [comment]
Great initiative!

One thing that should be considered is the token requirement for connect a SCIM client to a SCIM server.
Looking at some big players, like Azure Ad (EntraID now) or Okta, which provide a SCIM client, they both expect a long-live token and **do not** support any auth2 method to get and/or refresh the token.

Do we want to comply/implement the same way? That would raise some issues if we ever have a SCIM server in Keycloak, as it cannot provide (out-of-the-box) such token.
Do we want to user a "regular" auth2 token and differ from what the market seems to set as standard?

@Captain-P-Goldfish might have more insights about this, from a SCIM server point of view.

M.

**Comment by Captain-P-Goldfish** [comment]
> Great initiative!
> 
> One thing that should be considered is the token requirement for connect a SCIM client to a SCIM server. Looking at some big players, like Azure Ad (EntraID now) or Okta, which provide a SCIM client, they both expect a long-live token and **do not** support any auth2 method to get and/or refresh the token.
> 
> Do we want to comply/implement the same way? That would raise some issues if we ever have a SCIM server in Keycloak, as it cannot provide (out-of-the-box) such token. Do we want to user a "regular" auth2 token and differ from what the market seems to set as standard?
> 
> @Captain-P-Goldfish might have more insights about this, from a SCIM server point of view.
> 
> M.

Sorry, since I have created a little business around that feature I cannot give too much support on this. But I can tell you at least that just one authentication mechanism like long life-bearer-tokens is not enough.

**Comment by MartinEmrich** [comment]
Just lurking here, while using AWS IAM Identity Center against Entra ID...

Indeed I have to manually click to create a fresh SCIM token from Azure/Entra ID once a year, copy it, and paste it into AWS IAM Identity Center. 
AWS sends a reminder email well in advance, so it is not a big hassle IMHO.

**Comment by kfox1111** [comment]
Its currently unusable?

> © 2025 Metatavu. All rights reserved.

And no license file?

**Comment by kfox1111** [comment]
Its absolutely on topic. When the software reserves all rights, and there is no license provided, copyright/opensource respecting folks can not use it.

If we are to expect people to respect our own software licenses, we have to respect those of others.

What license is keycloak-scim-server under? If not stated, we shouldn't be looking at it as we don't even have the right granted to even look at it.

**Comment by anttileppa** [comment]
@kfox1111 Thanks for the reminder — you are absolutely correct. I had initially forgotten to add the license, but that has now been fixed. The project is under the same license as Keycloak itself: Apache License, Version 2.0.

I appreciate you highlighting this — respecting licenses is important, and I fully agree that we must hold ourselves to the same standards we expect from others.

**Comment by kfox1111** [comment]
Awesome. Thanks for clearing that up, and the very interesting project. :)

**Comment by lrozenblyum** [comment]
@anttileppa hi, your extension looks promising. You may promote it to the page https://www.keycloak.org/extensions by opening an issue on https://github.com/keycloak/keycloak-web/issues

**Comment by anttileppa** [comment]
@lrozenblyum I have opened a pull-request for adding the extension into that page: https://github.com/keycloak/keycloak-web/pull/612

**Comment by oculos** [comment]
It’s good to see so many ideas around scim and a Keycloak. 

I imagine different organizations have different needs. I started to write a small extension because my needs are very small and I would like to keep things modular:

- my implementation is mostly of a Scim client that sends info on login. 
- I am saving configuration on a Component Model tied to the client, since different clients might need different providers. I might change this and add it to the realm component model
tge inspiration is mostly to export groups to GitHub enterprise, but only when there is login events on the client using the scim configuration.

**Comment by zrav** [comment]
Looking forward to a native SCIM client to enable use cases like forwarding groups to Atlassian Cloud.

**Comment by cpesch** [comment]
You could give https://github.com/Metatavu/keycloak-scim-server a try

**Comment by pedroigor** [comment]
> You could give https://github.com/Metatavu/keycloak-scim-server a try

Definitely. We are looking at any extensions from the community to look for the capabilities we need to support. They are solving real-world use cases, and are very helpful in understanding the areas we should focus on.

Hopefully, this set of tasks is enough for the first release of this feature (Keycloak as  SCIM Provider, initially) to help you and other on integrating with Keycloak using SCIM for (de) provisioning of users:

* https://github.com/keycloak/keycloak/issues/45711
* https://github.com/keycloak/keycloak/issues/46213

**Comment by pedroigor** [comment]
Hi,

First of all, thanks for all the feedback. Even though it took some time, we have a first release of the SCIM API and a plan to make it supported. See https://www.keycloak.org/2026/04/scim-as-experimental-feature.

**Comment by InfoSec812** [comment]
Very exciting to see this come into the project! I know a lot of my customers have been asking for this for some time!

**Comment by pierreozoux** [comment]
Nice to see Keycloak becoming a scim service provider :)

What about keycloak becoming a scim client for other service providers?
(This is what we are still working on here - https://forge.libre.sh/libre.sh/keycloak-scim )

**Comment by pedroigor** [comment]
@pierreozoux, https://github.com/keycloak/keycloak/issues/43531 is planned right after we are done with the SCIM API and make it supported.

**Comment by pedroigor** [comment]
From a management PoV, what does it make sense to expose via Admin UI / API in terms of settings?

Right now, we are introducing support for mapping (any) user profile attribute to a SCIM custom schema. For instance, you can have a user profile attribute mapped to `urn:my:scim:extension:custom:User:myattribute` scim attribute.

From a service provider configuration, there are a few settings we can expose, but I'm unsure which ones make sense and their use cases. For instance, see https://datatracker.ietf.org/doc/html/rfc7643#page-44.

In regard to schemas, we can just leverage the existing UI/APIs from user profile (Realm Settings -> User Profile) so that an administrator can easily map SCIM attributes. For groups, we probably won't allow custom attributes as it would require using a similar mechanism as user profile, but for groups. Or is it a must for production deployments?

As for resource types, we won't be supporting custom resource types even though possible because of the `ScimResourceTypeSpi`.

Wdyt?

**Comment by amartin-maincare** [comment]
Hi @pedroigor 

> As for resource types, we won't be supporting custom resource types even though possible because of the ScimResourceTypeSpi.

From our perspective, the ability to define custom resource types would be essential for production adoption.

At my company, we currently use a proprietary format to manage users, credentials, clients, groups, and several custom-related resources. We’re planning to migrate to SCIM to align with an open standard, which would simplify integration and improve interoperability.

Here are a few key use cases driving this:

- We have internal custom applications used to configure and query clients, users, credentials, groups, and some custom related resources. These apps abstract away the complexity of the Keycloak Admin UI, limit access to specific features, and allow fine-grained control over what users can do. They also support custom fields and resource extensions.
- We also support third-party applications developed by our customers, which need to query information about users, groups, clients, and other domain-specific resources. 

While we’re moving toward SCIM for standardization, not supporting custom resource types would be a significant blocker for us.  Without this capability, we’d be forced to maintain multiple endpoints or fallback to our proprietary system—undermining the goal of having a unified, standards-based API. 

Ideally, we’d like a single SCIM endpoint that can manage not only standard resources (users, groups) but also our custom ones. Being able to define and expose custom resource types via SCIM would make this possible.

Would it be feasible to reconsider support for custom resource types through ScimResourceTypeSpi, or are there architectural constraints preventing this?

**Comment by heshank94** [comment]
Im doing a small POC using OneLogin and SCIM Provisioner that supports Keycloak. According to SCIM standards to update a user you can use below API URL format.

<domain>/scim/v2/Users/:user_id

But Keycloak expects this :user_id in the request body as an 'id'. But I have checked all the sources related to SCIM standards and this is not the expected behaviour. Hoping this will be resolved sooner in upcoming release.

**Comment by woprandi** [comment]
I feel the 26.7 release support is a SCIM server, not client.

**Comment by InfoSec812** [comment]
I was actually talking to @ahus1 at WeAreDevelopers in Berlin about similar issues today and I am planning to take a look at how these sorts of issues could be addressed going forward. I am about to go on summer holiday and plan to spend a few days looking into this and discussing in further detail with the Keycloak dev team in early August. Keep your feedback coming and I will attempt to ensure that we address these concerns during our forthcoming discussions.
